Parsimony/ParserBuilder/OptionBuilder.cs

```diff
@@ -0,0 +1,169 @@
+            }
+        }
+
+        /// <summary>
+        /// The long name of the option being built
+        /// </summary>
+        /// <exception cref="InvalidOperationException"> The value of <see cref="LongName"/> has already been set </exception>
+        internal OptionName.Long? LongName
+        {
+            get => _longName;
+            set
+            {
+                if (_longName != null) throw new InvalidOperationException();
+                _longName = value;
+            }
+        }
+
+        /// <summary>
+        /// The selector function used to extract or set values on the parent <typeparamref name="TOption"/>
+        /// </summary>
+        internal PropertySelector<TOption, TProp> Selector { get; }
+
+        /// <summary>
+        /// The parser function to be used to convert input string to the tyope of <typeparamref name="TProp"/>
+        /// </summary>
+        /// <remarks> 
+        /// If this is not set before construction the <see cref="TypeConverter"/> for the class will be reflected off 
+        /// and used. If that is incapable of conversion from string building the object will throw an exception
+        /// </remarks>
+        internal Func<string, TProp>? Parser { get; set; }
+
+
+
+        private OptionName.Short? _shortName;
+        private OptionName.Long? _longName;
+
+        //TODO:CN -- HelpBuilder
+
+
+        /// <summary>
+        /// Constructs a new option builder with the given names
+        /// </summary>
+        /// <param name="shortName"> a valid short name flag for the option </param>
+        /// <param name="longName"> a valid long name for the option </param>
+        /// <param name="selector"> The property selector that indicates the property this flag represents on the parent <typeparamref name="TOption"/> </param>
+        /// <exception cref="ArgumentException"> <paramref name="shortName"/> or <paramref name="longName"/> are invalid </exception>
+        /// <exception cref="ArgumentException"> <paramref name="selector"/> is not a valid <see cref="PropertySelector{TInput, TProp} "/></exception>
+        /// <exception cref="ArgumentNullException"> <paramref name="longName"/> or <paramref name="selector"/> are null </exception>
+        public OptionBuilder(char shortName, string longName, Expression<Func<TOption, TProp>> selector)
+        {
+            _shortName = OptionName.Parse(shortName.ToString()) as OptionName.Short;
+            _longName = OptionName.Parse(longName) as OptionName.Long;
+            Selector = new PropertySelector<TOption, TProp>(selector);
+        }
+
+        /// <summary>
+        /// Constructs a new option builder with the given short name
+        /// </summary>
+        /// <param name="selector"> The property selector that indicates the property this flag represents on the parent <typeparamref name="TOption"/> </param>
+        /// <param name="shortName"> a valid short name flag for the option </param>
```

Comment:
ive been using the terms "flag" and "option" to mean bool and non-bool. im not married to that particular use but we should try to be consistent in what ever we settle on

Reply:
Yeah I can fix that

Reply:
I think we should stop using TOption in that case for the parent viewmodel